File: app/src/main/java/com/ldp/reader/model/remote/RemoteRepository.java
package com.ldp.reader.model.remote;

import com.ldp.reader.model.bean.BookDetailBeanInOwn;
import com.ldp.reader.model.bean.BookIdBean;
import com.ldp.reader.model.bean.BookSearchResult;
import com.ldp.reader.model.bean.ChapterBean;
import com.ldp.reader.model.bean.ChapterInfoBean;
import com.ldp.reader.model.bean.ContentBean;
import com.ldp.reader.model.bean.DirectLoginResultBean;
import com.ldp.reader.model.bean.LoginResultBean;
import com.ldp.reader.model.bean.SmsLoginBean;
import com.ldp.reader.model.bean.SyncBookShelfBean;
import com.mob.secverify.datatype.VerifyResult;

import java.util.List;

import io.reactivex.Single;
import okhttp3.RequestBody;
import retrofit2.Retrofit;

import static com.ldp.reader.utils.Constant.APP_KEY;
import static com.ldp.reader.utils.Constant.APP_SECRET;

/**
 * Created by ldp on 17-4-20.
 */

public class RemoteRepository {
    private static RemoteRepository sInstance;
    private Retrofit mRetrofit, mRetrofitByOwn;
    private BookApi mBookApi;
    private BookApiOwn mBookApiOwn;


    private RemoteRepository() {
        mRetrofit = RemoteHelper.getInstance()
                .getRetrofit();
        mBookApi = mRetrofit.create(BookApi.class);


        mRetrofitByOwn = RemoteHelper.getInstance()
                .getRetrofitByOwn();
        mBookApiOwn = mRetrofitByOwn.create(BookApiOwn.class);
    }

    public static RemoteRepository getInstance() {
        if (sInstance == null) {
            synchronized (RemoteHelper.class) {
                if (sInstance == null) {
                    sInstance = new RemoteRepository();
                }
            }
        }
        return sInstance;
    }



    public Single<List<BookSearchResult>> getSearchResult(String bookName) {
        return mBookApiOwn.getSearchResult(bookName);

    }


    public Single<BookDetailBeanInOwn> getBookInfo(String bookId) {
        return mBookApiOwn.getBookInfo(bookId);
    }

    public Single<List<BookDetailBeanInOwn>> getBookInfoBatch( RequestBody body) {
        return mBookApiOwn.getBookInfoBatch(body);
    }



    public Single<List<ChapterBean>> getBookFolder(String bookId) {
        return mBookApiOwn.getBookFolder(bookId);
    }

    public Single<ContentBean> getBookContent(String bookId,String chapterId ,int sourceIndex) {
        return mBookApiOwn.getBookContent(bookId,chapterId,sourceIndex);
    }

    public Single<LoginResultBean> userLogin(String userNameInput, String passwordInput) {
        return mBookApiOwn.userLogin(userNameInput,passwordInput);
    }

    public Single<SmsLoginBean> smsLogin(String phoneNumber , String smsCode , String registrationId) {
        return mBookApiOwn.smsLogin( phoneNumber , smsCode , registrationId) ;
    }

    public Single<DirectLoginResultBean> userDirectLogin(VerifyResult verifyResult,String registrationId) {
        return mBookApiOwn.userDirectLogin(APP_KEY,APP_SECRET,verifyResult.getToken() ,verifyResult.getOpToken(),verifyResult.getOperator(),registrationId) ;
    }

    public Single<List<BookIdBean>> getBookShelf(String header) {
        return mBookApiOwn.getBookShelf(header);
    }

    public Single<List<BookIdBean>> getBookShelfByMobile(String mobile ,String token) {
        return mBookApiOwn.getBookShelfByMobile(mobile,token);
    }

    public Single<SyncBookShelfBean> setBookShelf(String token, RequestBody body) {
        return mBookApiOwn.setBookShelf(token,body);
    }


    public Single<SyncBookShelfBean> setBookShelfByMobile(RequestBody body) {
        return mBookApiOwn.setBookShelfByMobile(body);
    }


    /**
     * 注意这里用的是同步请求
     *
     * @param url
     * @return
     */
    public Single<ChapterInfoBean> getChapterInfo(String url) {
        return mBookApi.getChapterInfoPackage(url)
                .map(bean -> bean.getChapter());
    }

    /********************************书籍搜索*********************************************/
    /**
     * 搜索热词
     *
     * @return
     */
    public Single<List<String>> getHotWords() {
        return mBookApi.getHotWordPackage()
                .map(bean -> bean.getHotWords());
    }

    /**
     * 搜索关键字
     *
     * @param query
     * @return
     */
    public Single<List<String>> getKeyWords(String query) {
        return mBookApi.getKeyWordPacakge(query)
                .map(bean -> bean.getKeywords());

    }



}

File: app/src/main/java/com/ldp/reader/model/remote/RemoteRepository.kt
package com.ldp.reader.model.remote

import com.ldp.reader.model.bean.BookDetailBeanInOwn
import com.ldp.reader.model.bean.BookIdBean
import com.ldp.reader.model.bean.BookSearchResult
import com.ldp.reader.model.bean.ChapterBean
import com.ldp.reader.model.bean.ContentBean
import com.ldp.reader.model.bean.DirectLoginResultBean
import com.ldp.reader.model.bean.LoginResultBean
import com.ldp.reader.model.bean.SmsLoginBean
import com.ldp.reader.model.bean.SyncBookShelfBean
import com.ldp.reader.source.SourceNetworkDispatchers
import com.ldp.reader.source.SourceNetworkForegroundPriority
import com.ldp.reader.utils.Constant.APP_KEY
import com.ldp.reader.utils.Constant.APP_SECRET
import com.mob.secverify.datatype.VerifyResult
import kotlinx.coroutines.withContext
import okhttp3.RequestBody
import retrofit2.Call
import retrofit2.HttpException
import retrofit2.Retrofit

class RemoteRepository private constructor() {
    private val mRetrofit: Retrofit
    private val mRetrofitByOwn: Retrofit
    private val mBookApi: BookApi
    private val mBookApiOwn: BookApiOwn

    init {
        mRetrofit = RemoteHelper.getInstance().getRetrofit()
        mBookApi = mRetrofit.create(BookApi::class.java)

        mRetrofitByOwn = RemoteHelper.getInstance().getRetrofitByOwn()
        mBookApiOwn = mRetrofitByOwn.create(BookApiOwn::class.java)
    }

    suspend fun getSearchResult(bookName: String?): List<BookSearchResult> {
        return execute(mBookApiOwn.getSearchResult(bookName))
    }

    suspend fun getBookInfo(bookId: String?): BookDetailBeanInOwn {
        return execute(mBookApiOwn.getBookInfo(bookId))
    }

    suspend fun getBookInfoBatch(body: RequestBody?): List<BookDetailBeanInOwn> {
        return execute(mBookApiOwn.getBookInfoBatch(body))
    }

    suspend fun getBookFolder(bookId: String?): List<ChapterBean> {
        return execute(mBookApiOwn.getBookFolder(bookId))
    }

    suspend fun getBookContent(bookId: String?, chapterId: String?, sourceIndex: Int): ContentBean {
        return execute(mBookApiOwn.getBookContent(bookId, chapterId, sourceIndex))
    }

    suspend fun userLogin(userNameInput: String?, passwordInput: String?): LoginResultBean {
        return execute(mBookApiOwn.userLogin(userNameInput, passwordInput))
    }

    suspend fun smsLogin(phoneNumber: String?, smsCode: String?, registrationId: String?): SmsLoginBean {
        return execute(mBookApiOwn.smsLogin(phoneNumber, smsCode, registrationId))
    }

    suspend fun userDirectLogin(verifyResult: VerifyResult?, registrationId: String?): DirectLoginResultBean {
        return execute(mBookApiOwn.userDirectLogin(
            APP_KEY,
            APP_SECRET,
            verifyResult!!.token,
            verifyResult.opToken,
            verifyResult.operator,
            registrationId
        ))
    }

    suspend fun getBookShelf(header: String?): List<BookIdBean> {
        return execute(mBookApiOwn.getBookShelf(header))
    }

    suspend fun getBookShelfByMobile(mobile: String?, token: String?): List<BookIdBean> {
        return execute(mBookApiOwn.getBookShelfByMobile(mobile, token))
    }

    suspend fun setBookShelf(token: String?, body: RequestBody?): SyncBookShelfBean {
        return execute(mBookApiOwn.setBookShelf(token, body))
    }

    suspend fun setBookShelfByMobile(body: RequestBody?): SyncBookShelfBean {
        return execute(mBookApiOwn.setBookShelfByMobile(body))
    }

    suspend fun getHotWords(): List<String> {
        return execute(mBookApi.getHotWordPackage()).hotWords!!
    }

    suspend fun getKeyWords(query: String?): List<String> {
        return execute(mBookApi.getKeyWordPacakge(query)).keywords!!
    }

    private suspend fun <T> execute(call: Call<T>): T {
        val request = call.request()
        return SourceNetworkForegroundPriority.entered(
            operation = "remote",
            key = "${request.method} ${request.url.encodedPath}"
        ) {
            withContext(SourceNetworkDispatchers.foreground) {
                val response = call.execute()
                if (!response.isSuccessful) {
                    throw HttpException(response)
                }
                response.body()!!
            }
        }
    }

    companion object {
        @Volatile
        private var sInstance: RemoteRepository? = null

        @JvmStatic
        fun getInstance(): RemoteRepository {
            if (sInstance == null) {
                synchronized(RemoteHelper::class.java) {
                    if (sInstance == null) {
                        sInstance = RemoteRepository()
                    }
                }
            }
            return sInstance!!
        }
    }
}
